server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java

```diff
@@ -45,43 +52,154 @@
+        MimeMessage message = null;
+        Map<String, Serializable> attributes = ImmutableMap.<String, Serializable> of();
+        FakeMailContext.SentMail expected = new FakeMailContext.SentMail(sender, recipients, message, attributes);
         assertThat(fakeMailContext.getSentMails()).containsOnly(expected);
     }
 
+    @Test
+    public void bounceShouldChangeTheStateWhenNoSenderAndPassThroughEqualsFalse() throws Exception {
+        FakeMailetConfig mailetConfig = FakeMailetConfig.builder()
+                .mailetName(MAILET_NAME)
+                .mailetContext(fakeMailContext)
+                .setProperty("passThrough", "false")
+                .build();
+        bounce.init(mailetConfig);
+
+        FakeMail mail = FakeMail.builder()
+                .name(MAILET_NAME)
+                .recipient(MailAddressFixture.ANY_AT_JAMES)
+                .build();
+
+        MimeMessage mimeMessage = new MimeMessage(Session.getDefaultInstance(new Properties()));
+        mimeMessage.setText("My content");
+        mail.setMessage(mimeMessage);
+
+        bounce.service(mail);
+
+        assertThat(mail.getState()).isEqualTo(Mail.GHOST);
+    }
+
+    @Test
+    public void bounceShouldNotChangeTheStateWhenNoSenderAndPassThroughEqualsTrue() throws Exception {
+        FakeMailetConfig mailetConfig = FakeMailetConfig.builder()
+                .mailetName(MAILET_NAME)
+                .mailetContext(fakeMailContext)
+                .setProperty("passThrough", "true")
+                .build();
+        bounce.init(mailetConfig);
+
+        String initialState = "initial";
+        FakeMail mail = FakeMail.builder()
+                .state(initialState)
+                .name(MAILET_NAME)
+                .recipient(MailAddressFixture.ANY_AT_JAMES)
+                .build();
+
+        MimeMessage mimeMessage = new MimeMessage(Session.getDefaultInstance(new Properties()));
+        mimeMessage.setText("My content");
+        mail.setMessage(mimeMessage);
+
+        bounce.service(mail);
+
+        assertThat(mail.getState()).isEqualTo(initialState);
+    }
+
+    @Test
+    public void bounceShouldAddPrefixToSubjectWhenPrefixIsConfigured() throws Exception {
+        FakeMailetConfig mailetConfig = FakeMailetConfig.builder()
+                .mailetName(MAILET_NAME)
+                .mailetContext(fakeMailContext)
+                .setProperty("prefix", "pre")
+                .build();
+        bounce.init(mailetConfig);
+
+        MimeMessage mimeMessage = new MimeMessage(Session.getDefaultInstance(new Properties()));
+        mimeMessage.setSubject("My subject");
+        FakeMail mail = FakeMail.builder()
+                .name(MAILET_NAME)
+                .sender(MailAddressFixture.ANY_AT_JAMES)
+                .mimeMessage(mimeMessage)
+                .build();
+
+        bounce.service(mail);
+
+        assertThat(mail.getMessage().getSubject()).isEqualTo("pre My subject");
```

Comment:
Same test with such configuration, and not so simple to achieve.
Testing sent mail should be an integration test.